### PR TITLE
Fixes #24550 - Search repo by products

### PR DESCRIPTION
**Pull request by waldenraines**
Extended https://github.com/Katello/katello/pull/7702 and fixed the placement of the Export as CSV button.

### CODE REVIEW

**Comment by johnpmitsch** [comment]
@waldenraines what is the reasoning behind putting redux files in `webpack/redux`? I thought we were moving away from that using `scenes/`. Also, there is a `webpack/redux/reducers` but `OrganizationProductsReducer.js` is not in there, so seems like we are introducing a different pattern even within the folder.

**Comment by waldenraines** [comment]
> @waldenraines what is the reasoning behind putting redux files in `webpack/redux`? I thought we were moving away from that using `scenes/`. Also, there is a `webpack/redux/reducers` but `OrganizationProductsReducer.js` is not in there, so seems like we are introducing a different pattern even within the folder.

Yeah, we are sort of between patterns here.  Everything from the repositories pages needs to be moved into the new structure but I think @sharvit didn't want to take that on in this fix since it's just a bug fix.

**Comment by johnpmitsch** [comment]
> To me it doesn't really matter that much, we know we need to move it and a lot of things will move around and be renamed etc. I will likely be the one to do this and I think it's fine.

I'm ok with this as long as it happens in a reasonable amount of time. We now have a lot of conflicting patterns in this area and it makes it hard for other devs who are new to React to make sense of things

File: app/models/katello/product_content.rb [hunk 1]
```diff
@@ -22,6 +22,7 @@ class ProductContent < Katello::Model
     scoped_search :on => :content_type, :relation => :content
     scoped_search :on => :label, :relation => :content
     scoped_search :on => :name, :relation => :product, :rename => :product_name
+    scoped_search :on => :id, :relation => :product, :rename => :product_id
```
Comment (jlsherrill) [inline app/models/katello/product_content.rb]: can you add ':only_explicit => true'  we typically do that for ids

File: app/models/katello/repository.rb [hunk 2]
```diff
@@ -183,6 +183,8 @@ class Repository < Katello::Model
     scoped_search :on => :redhat, :complete_value => { :true => true, :false => false }, :ext_method => :search_by_redhat
     scoped_search :on => :container_repository_name, :complete_value => true
     scoped_search :on => :description, :only_explicit => true
+    scoped_search :on => :name, :relation => :product, :rename => :product_name
+    scoped_search :on => :id, :relation => :product, :rename => :product_id
```
Comment (jlsherrill) [inline app/models/katello/repository.rb]: can you add ':only_explicit => true' we typically do that for ids

**Comment by sharvit** [comment]
Thanks for extending my work @waldenraines :+1: 

> > @waldenraines what is the reasoning behind putting redux files in `webpack/redux`? I thought we were moving away from that using `scenes/`. Also, there is a `webpack/redux/reducers` but `OrganizationProductsReducer.js` is not in there, so seems like we are introducing a different pattern even within the folder.
> 
> Yeah, we are sort of between patterns here. Everything from the repositories pages needs to be moved into the new structure but I think @sharvit didn't want to take that on in this fix since it's just a bug fix.

Actually, I had another reason to do so.
I think we are abusing the `scenes/` folder. It made to contain the different pages hierarchically with their data mechanism (actions, reducers, selectors...).

Should we also use the `scenes/` folder when we need a shared data endpoint with multiple consumers and no main component?
IMHO we should have `scenes/` and `redux/` living side by side.

For example, IMO those endpoint shouldn't be in the `scenes/` folder:
https://github.com/Katello/katello/tree/master/webpack/scenes/Organizations
https://github.com/Katello/katello/tree/master/webpack/scenes/Products
https://github.com/Katello/katello/tree/master/webpack/scenes/Settings
https://github.com/Katello/katello/tree/master/webpack/scenes/Tasks

**Comment by johnpmitsch** [comment]
> > Yeah, we are sort of between patterns here. Everything from the repositories pages needs to be moved into the new structure but I think @sharvit didn't want to take that on in this fix since it's just a bug fix.
> 
> Actually, I had another reason to do so.
> I think we are abusing the `scenes/` folder. It made to contain the different pages hierarchically with their data mechanism (actions, reducers, selectors...).

You may have some valid points around where we are putting things, but I don't think the answer is to introduce more patterns in PRs. I'm assuming the goal of putting things in `scenes/` was to get on a consistent file structure and pattern, and last I heard we were moving most files in `redux/` to get this consistency.

I think having a consistent folder structure is *way* more important than what the folder structure is. Introducing new patterns based on an opinion that hasn't been discussed doesn't seem like the right way to go about things, this makes it hard for other developers to understand this area of code, which is already new to those outside the UX team.

**Comment by akofink** [comment]
> I think having a consistent folder structure is way more important than what the folder structure is

Agreed. This PR isn't the place to change our React/Redux structure, but I think we should change all of it to match Foreman's structure in a PR in the future. It'd be good to try and figure out what that'd cost in terms of developer time.

**Comment by waldenraines** [comment]
@sharvit let me know when you have a new PR for this and I will close.

**Comment by waldenraines** [comment]
> * Don't show custom products in the list of products.

Not currently possible with the API.  If I provide custom=false I still get custom products, see:  https://github.com/Katello/katello/blob/b7dc5d44d9d2af3b64a0c063a67f1e810be891cb/app/controllers/katello/api/v2/products_controller.rb#L50

> * We should enable search on this dropdown in a future PR. There can be way too many products to just pick one from the list (I struggled to find RHEL 7 Server product). @waldenraines would you mind filing an issue for that?

https://projects.theforeman.org/issues/25171

> * Content type no longer defaults to RPM - is this an intended change? I see now that this changed before this PR. :)

Fixed with https://github.com/Katello/katello/pull/7733

**Comment by waldenraines** [comment]
> Is there a way to limit the scoped search results to the products selected? This would really help to find specific repos. i.e. if I have "Red Hat Satellite" selected and then type 'name =', its shows the names of the repos only inside of 'Red Hat Satellite' product. Not sure if this is limited by scoped search. 

I'm not sure that the autocomplete search supports this.  If you search for `product_name` in the search box and then for `name` the name results are not filtered by `product_name`.  Hopefully I explained that well.

> Is there a reason to keep the search bar to half the page width? I don't see anything on the other side. I could see users creating some large search queries on this page, so it would be nice to have the space if its there.

It's half the width because it's following the patternfly pattern.  If we wanted we could increase the size of it.  @rohoover thoughts?

> A way of clearning all the selections on the product dropdown would be nice.

I could look into wiring that into the 'x'

> With the exception of custom products being shown, I'm ok with follow-up issues and PRs. Just sharing my initial UX impressions.

As stated above this isn't possible without an API change as far as I can tell.

**Comment by johnpmitsch** [comment]
> Not currently possible with the API. If I provide custom=false I still get custom products, see:

If I'm understanding this correctly, it sounds like a bug. Can we file an issue?

**Comment by waldenraines** [comment]
> If I'm understanding this correctly, it sounds like a bug. Can we file an issue?

Sure, go for it :stuck_out_tongue_closed_eyes: 

JK, here you go: https://projects.theforeman.org/issues/25180

**Comment by waldenraines** [comment]
> > A way of clearning all the selections on the product dropdown would be nice.
> 
> I could look into wiring that into the 'x'

@johnpmitsch do you mind if we address this as part of https://projects.theforeman.org/issues/25171?

**Comment by waldenraines** [comment]
@johnpmitsch @akofink any other thoughts?

**Comment by waldenraines** [comment]
[test katello]

**Comment by akofink** [comment]
Also, we should fix the errors on this page. Not necessarily in this PR (unless caused by it), but we should at least file issues:

```
checkPropTypes.js:19 Warning: Failed prop type: The prop `pagination.page` is marked as required in `PaginationRow`, but its value is `undefined`.
    in PaginationRow (created by RedHatRepositoriesPage)
    in RedHatRepositoriesPage (created by Connect(RedHatRepositoriesPage))
    in Connect(RedHatRepositoriesPage) (created by Headers)
    in Headers (created by CheckOrg)
    in CheckOrg (created by Route)
    in Route
    in div
    in Unknown (created by Application)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by Application)
    in Application (created by Connect(Application))
    in Connect(Application)
    in Provider

Warning: Failed prop type: The prop `itemCount` is marked as required in `PaginationRow`, but its value is `undefined`.
    in PaginationRow (created by RedHatRepositoriesPage)
    in RedHatRepositoriesPage (created by Connect(RedHatRepositoriesPage))
    in Connect(RedHatRepositoriesPage) (created by Headers)
    in Headers (created by CheckOrg)
    in CheckOrg (created by Route)
    in Route
    in div
    in Unknown (created by Application)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by Application)
    in Application (created by Connect(Application))
    in Connect(Application)
    in Provider

Warning: Failed prop type: Invalid prop `maxItemsCountForFullLabel` of type `string` supplied to `BootstrapSelect`, expected `number`.
    in BootstrapSelect (created by MultiSelect)
    in MultiSelect (created by SearchBar)
    in div (created by SearchBar)
    in form (created by Form)
    in Form (created by SearchBar)
    in SearchBar (created by Connect(SearchBar))
    in Connect(SearchBar) (created by RedHatRepositoriesPage)
    in div (created by Col)
    in Col (created by RedHatRepositoriesPage)
    in div (created by Row)
    in Row (created by RedHatRepositoriesPage)
    in div (created by Grid)
    in Grid (created by RedHatRepositoriesPage)
    in RedHatRepositoriesPage (created by Connect(RedHatRepositoriesPage))
    in Connect(RedHatRepositoriesPage) (created by Headers)
    in Headers (created by CheckOrg)
    in CheckOrg (created by Route)
    in Route
    in div
    in Unknown (created by Application)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by Application)
    in Application (created by Connect(Application))
    in Connect(Application)
    in Provider

Warning: The `defaultValue` prop supplied to <select> must be an array if `multiple` is true.

Check the render method of `FormControl`.
    in select (created by FormControl)
    in FormControl (created by BootstrapSelect)
    in BootstrapSelect (created by MultiSelect)
    in div (created by FormGroup)
    in FormGroup (created by MultiSelect)
    in MultiSelect (created by SearchBar)
    in div (created by SearchBar)
    in form (created by Form)
    in Form (created by SearchBar)
    in SearchBar (created by Connect(SearchBar))
    in Connect(SearchBar) (created by RedHatRepositoriesPage)
    in div (created by Col)
    in Col (created by RedHatRepositoriesPage)
    in div (created by Row)
    in Row (created by RedHatRepositoriesPage)
    in div (created by Grid)
    in Grid (created by RedHatRepositoriesPage)
    in RedHatRepositoriesPage (created by Connect(RedHatRepositoriesPage))
    in Connect(RedHatRepositoriesPage) (created by Headers)
    in Headers (created by CheckOrg)
    in CheckOrg (created by Route)
    in Route
    in div
    in Unknown (created by Application)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by Application)
    in Application (created by Connect(Application))
    in Connect(Application)
    in Provider
```

Hopefully I'm not the only one seeing these. Can you confirm?

**Comment by waldenraines** [comment]
> These are verbatim comments from my previous review. Neither seem resolved?

I was wondering if we could address those in a future PR as mentioned above.

> Search for products in dropdown
> * I think there is a follow up issue, but can't find it. We should file this if it doesn't exist

https://projects.theforeman.org/issues/25171

> * We are introducing a new pattern that conflicts with existing ones. I would much rather we stick to the organization patterns that currently exist, then change them in a larger refactoring if its desired. I know there is a plan to change things, but often times these cosmetic refactorings get delayed or never happen and we have to live with the "temporary" solution permanently.

I don't know what to say here, @sharvit do you want to respond to this?

> * Search supporting only autocompleting the names from the products selected.
>   
>   * I think there are some technical limitations here with scoped search, but its worth investigating. We possibly could set up a new autocomplete endpoint and/or do two queries. I think this would be a huge improvement to this page's UX.

 I agree but the necessary ruby changes are beyond this team's expertise.  I would definitely invite a collaboration on this.

> * Comparing this to master, it looks like there are two new JS errors. This PR shouldn't introduce new errors. At a minimum, these should be fixed. If the other two are easy to fix, please fix them as well or file an issue and we can address later

I'll look into fixing all of the JS errors on this page as part of this PR.

**Comment by waldenraines** [comment]
@johnpmitsch @akofink fixed all of the JS errors on the page.

**Pull request merged**